### PR TITLE
 boot: helper for generating secboot load chains from a given boot asset sequence

### PR DESCRIPTION
**Pull request by bboozzoo**
The helpers enables generating secboot load chain trees from boot asset
sequences. 

This is stacked on top of #9303. The relevant commit is https://github.com/snapcore/snapd/commit/687fa3bb505b2813dc90bf326ecac476cb245839

### CODE REVIEW

File: boot/assets.go [hunk 1]
```diff
@@ -61,6 +57,10 @@ func (c *trustedAssetsCache) pathInCache(part string) string {
 	return filepath.Join(c.cacheDir, part)
 }
 
+func trustedAssetCacheKey(blName, assetName, assetHash string) string {
```
Comment (pedronis) [inline boot/assets.go:60]: given the usage maybe this should be called s/Key/RelPath/

File: boot/assets.go [hunk 2]
```diff
@@ -49,10 +49,6 @@ func newTrustedAssetsCache(cacheDir string) *trustedAssetsCache {
 	return &trustedAssetsCache{cacheDir: cacheDir, hash: crypto.SHA3_384}
 }
 
-func (c *trustedAssetsCache) assetKey(blName, assetName, assetHash string) string {
-	return filepath.Join(blName, fmt.Sprintf("%s-%s", assetName, assetHash))
-}
-
 func (c *trustedAssetsCache) tempAssetKey(blName, assetName string) string {
```
Comment (pedronis) [inline boot/assets.go:52]: same s/Key/RelPath/ then for consistency

File: boot/bootchain.go [hunk 3]
```diff
@@ -195,3 +202,44 @@ func predictableBootChainsEqualForReseal(pb1, pb2 predictableBootChains) bool {
...
+		p := filepath.Join(
+			dirs.SnapBootAssetsDir,
+			trustedAssetCacheKey(blName, thisAsset.Name, hash))
+		if !osutil.FileExists(p) {
+			return nil, fmt.Errorf("file %s not found in assets cache", p)
+		}
+		bf = bootloader.NewBootFile(
+			"", // asset comes from the filesystem, not a snap
+			p,
+			bootloader.Role(thisAsset.Role),
```
Comment (pedronis) [inline boot/bootchain.go:236]: the json package docs are not very transparent on this, but we can make bootAsset Role a bootloader.Role and avoid some of the back and forth at the cost of changing the test, should be done on its own though

File: boot/bootchain.go [hunk 4]
```diff
@@ -195,3 +202,44 @@ func predictableBootChainsEqualForReseal(pb1, pb2 predictableBootChains) bool {
...
+	// kernel is added after all the assets
+	addKernelBootFile := len(assets) == 0
+	if addKernelBootFile {
+		return []*secboot.LoadChain{secboot.NewLoadChain(kernelBootFile)}, nil
+	}
+
+	thisAsset := assets[0]
+	blName := roleToBlName[thisAsset.Role]
+	if blName == "" {
+		return nil, fmt.Errorf("internal error: no bootloader name for asset role %q", thisAsset.Role)
```
Comment (pedronis) [inline boot/bootchain.go:219]: s/asset/boot asset/

I think we should make a pass to make sure all errors using "asset" have either bootloader or boot mentioned in it

File: boot/bootchain.go [hunk 5]
```diff
@@ -195,3 +202,44 @@ func predictableBootChainsEqualForReseal(pb1, pb2 predictableBootChains) bool {
...
+	for _, hash := range thisAsset.Hashes {
+		var bf bootloader.BootFile
+		var next []*secboot.LoadChain
+		var err error
+
+		p := filepath.Join(
+			dirs.SnapBootAssetsDir,
+			trustedAssetCacheKey(blName, thisAsset.Name, hash))
+		if !osutil.FileExists(p) {
+			return nil, fmt.Errorf("file %s not found in assets cache", p)
```
Comment (pedronis) [inline boot/bootchain.go:231]: boot assets cache

**Review by pedronis — APPROVED** [review]
thank you, looks good, some detail comments

File: boot/bootchain.go [hunk 6]
```diff
@@ -48,9 +48,9 @@ type bootChain struct {
 
 // TODO:UC20 add a doc comment when this is stabilized
 type bootAsset struct {
-	Role   string   `json:"role"`
-	Name   string   `json:"name"`
-	Hashes []string `json:"hashes"`
+	Role   bootloader.Role `json:"role"`
```
Comment (pedronis) [inline boot/bootchain.go:51]: interesting, this was easier than I thought, but good

File: boot/bootchain_test.go [hunk 7]
```diff
@@ -987,3 +1004,169 @@ func (s *sealSuite) TestPredictableBootChainsSortOrder(c *C) {
...
+	})
+}
+
+func (s *bootchainSuite) TestBootAssetsToLoadChainErr(c *C) {
+	kbl := bootloader.NewBootFile("pc-kernel", "kernel.efi", bootloader.Role("run"))
+
+	assets := []boot.BootAsset{
+		{Name: "shim", Hashes: []string{"hash0"}, Role: "recovery"},
+		{Name: "loader-recovery", Hashes: []string{"hash0"}, Role: "recovery"},
+		{Name: "loader-run", Hashes: []string{"hash0"}, Role: "run"},
```
Comment (cmatsuoka) [inline boot/bootchain_test.go:1042]: It doesn't make any difference here, but the run mode Role as defined in bootloader is "run-mode".

**Review by cmatsuoka — APPROVED** [review]
Thanks Maciek, lgtm.

**Pull request merged**